gazebo_plugins/src/gazebo_ros_joint_state_publisher.cpp

```diff
@@ -0,0 +1,183 @@
+// are met:
+//
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above
+//    copyright notice, this list of conditions and the following
+//    disclaimer in the documentation and/or other materials provided
+//    with the distribution.
+//  * Neither the name of the {company} nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+// COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include <gazebo/common/Events.hh>
+#include <gazebo/common/Time.hh>
+#include <gazebo/physics/Joint.hh>
+#include <gazebo/physics/Model.hh>
+#include <gazebo/physics/World.hh>
+#include <gazebo_plugins/gazebo_ros_joint_state_publisher.hpp>
+#include <gazebo_ros/conversions.hpp>
+#include <gazebo_ros/node.hpp>
+#include <rclcpp/rclcpp.hpp>
+#include <sensor_msgs/msg/joint_state.hpp>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace gazebo_plugins
+{
+class GazeboRosJointStatePublisherPrivate
+{
+public:
+  /// Callback to be called at every simulation iteration.
+  /// \param[in] info Updated simulation info.
+  void OnUpdate(const gazebo::common::UpdateInfo & info);
+
+  /// A pointer to the GazeboROS node.
+  gazebo_ros::Node::SharedPtr ros_node_;
+
+  /// Joint state publisher.
+  rclcpp::Publisher<sensor_msgs::msg::JointState>::SharedPtr joint_state_pub_;
+
+  /// Joints being tracked.
+  std::vector<gazebo::physics::JointPtr> joints_;
+
+  /// Period in seconds
+  double update_period_;
+
+  /// Keep last time an update was published
+  gazebo::common::Time last_update_time_;
+
+  /// Pointer to the update event connection.
+  gazebo::event::ConnectionPtr update_connection_;
+};
+
+GazeboRosJointStatePublisher::GazeboRosJointStatePublisher()
+: impl_(std::make_unique<GazeboRosJointStatePublisherPrivate>())
+{
+}
+
+GazeboRosJointStatePublisher::~GazeboRosJointStatePublisher()
+{
+}
+
+void GazeboRosJointStatePublisher::Load(gazebo::physics::ModelPtr model, sdf::ElementPtr sdf)
+{
+  auto logger = rclcpp::get_logger("gazebo_ros_force");
```

Comment:
Should be ```"gazebo_ros_joint_state_publisher"```, though I personally favor creating the Node in the beginning and using it's logger ```node->get_logger()```. This will avoid any inconsistency in logging if the node name is remapped

Reply:
I actually started that way, but then I moved the node creation to the bottom so we don't need to destroy the node in case the load function returns early.

But I can move it back to the top if you have a strong preference, I agree that it would make the logging consistent.